crates/core/tedge/src/cli/mqtt/subscribe.rs

```diff
@@ -90,75 +93,100 @@ fn subscribe(cmd: &MqttSubscribeCommand) -> Result<(), anyhow::Error> {
         options.set_transport(rumqttc::Transport::tls_with_config(tls_config.into()));
     }
 
-    match cmd.duration {
-        Some(Duration::ZERO) | None => {}
-        Some(duration) if duration.as_secs() < 5 => {
-            options.set_keep_alive(duration);
-        }
-        Some(_) => {
-            options.set_keep_alive(Duration::from_secs(5));
-        }
-    }
-
     let (client, mut connection) = Client::new(options, DEFAULT_QUEUE_CAPACITY);
     let interrupted = super::disconnect_if_interrupted(client.clone());
-    let started = std::time::Instant::now();
     let mut n_packets = 0;
 
-    for event in connection.iter() {
-        match event {
-            Ok(Event::Incoming(Packet::Publish(message))) => {
-                // trims the trailing null char if one exists
-                let payload = message
-                    .payload
-                    .strip_suffix(&[0])
-                    .unwrap_or(&message.payload);
-                match std::str::from_utf8(payload) {
-                    Ok(payload) => {
-                        if cmd.hide_topic {
-                            println!("{}", &payload);
-                        } else {
-                            println!("[{}] {}", &message.topic, payload);
-                        }
-                        n_packets += 1;
-                        if matches!(cmd.count, Some(count) if count > 0 && n_packets >= count) {
-                            eprintln!("INFO: Break");
-                            break;
-                        }
-                    }
-                    Err(err) => error!("{err}"),
+    match cmd.duration {
+        None => {
+            for event in connection.iter() {
+                if !process_event(
+                    event,
+                    client.clone(),
+                    cmd,
+                    interrupted.clone(),
+                    &mut n_packets,
+                ) {
+                    break;
                 }
             }
-            Ok(Event::Outgoing(Outgoing::PingReq)) => {
-                if matches!(cmd.duration, Some(duration) if !duration.is_zero() && started.elapsed() > duration)
-                {
+        }
+        Some(duration) => loop {
+            match connection.recv_timeout(duration) {
```

Comment:
This is does not behave as you expect.

Instead of waiting at most the given `duration` between *two* events, you want the *loop* to run at most the given `duration`.

```suggestion
        Some(duration) => while started.elapsed() < duration {
            match connection.recv_timeout(duration - started.elapsed() ) {
```